clippy_lints/src/borrow_deref_ref.rs

```diff
@@ -73,6 +73,9 @@ impl<'tcx> LateLintPass<'tcx> for BorrowDerefRef {
                 }
             })
             && !is_from_proc_macro(cx, e)
+            && let e_ty = cx.typeck_results().expr_ty_adjusted(e)
+            // check if the reference is coercing to a mutable reference
+            && !matches!(e_ty.kind(), ty::Ref(_, _, Mutability::Mut))
```

Comment:
I think that if the deref target is mutable, the lint should trigger:

```suggestion
            && !(matches!(e_ty.kind(), ty::Ref(_, _, Mutability::Mut)) && !is_mutable(cx, deref_target))
```

This should help you lint things such as:

```rust
    (&*{s}).foo();
```

or like

```rust
    let mut s = &Struct;
    s.foo(); // To avoid a warning about `s` not needed to be mutable
    (&*s).foo();
```

Reply:
Thanks!